### PR TITLE
Redesign BatchedFactory

### LINKED ISSUE #317: Batch circuit executions in BatchedFactories
To speed up runs when running on, e.g., IBMQ, all scaled circuits should be generated and then sent as a single job to execute. As opposed to scaling a single circuit then single a single circuit as a job to execute.

### PR DESCRIPTION
**Pull request by rmlarose**
WIP. Partially fixes #317.

@andreamari almost all tests of `BatchedFactory`s use `iterate`. The redesign gets rid of `iterate` because it doesn't make sense to iterate anymore if we send all jobs in one batch. Does this seem reasonable? If so I'll change these tests. 

Other feedback welcome.

To completely fix #317, we'll have to allow for a "batched executor."

### CODE REVIEW

**Comment by rmlarose** [comment]
NB: Also restructures inference.py into a module as this file was pretty long.

**Comment by rmlarose** [comment]
Ready for "pre-review" and discussion before fixing tests (most failing tests are caused by `Factory` -> `BaseFactory`). What do you think of the structure @andreamari @karalekas @nathanshammah @willzeng?

The biggest change is `BatchedFactory` now defines a [`run_batched`](https://github.com/unitaryfund/mitiq/pull/364/files#diff-3107e5b9944f377f85584de86ca20b13R182) method.

**Review by andreamari — COMMENTED** [review]
Thanks @rmlarose the overall structure looks good to me!

My comments:

- Up to now  `_instack` and `_outstack` where used only to save the data which have been actually measured by the executor not the initialization parameters. This means that when a factory is initialized they are empty, while they become simultaneously populated when running the factory.  Now instead `_instack ` is directly initialized by `__init__`  which means that, before running the factory, we have an intermediate situation in which `_instack` is full while `_outstack` is empty. 
If you think this is ok, we can leave the current structure. Otherwise an alternative option could be to define a private a method `_update_instack()` and to call it only when the measured expectation values are saved into `_outstack`. I have no strong preference here, just to double checking we are aware of this change of "unwritten convention".

- Shall we rename `iterate` to `run_classical` also in AdaptiveFactory? (To have a more uniform names).

- I left two minor code suggestions about `kwargs`.

**Comment by rmlarose** [comment]
Thanks @andreamari. 
> * Up to now  `_instack` and `_outstack` where used only to save the data which have been actually measured by the executor not the initialization parameters. This means that when a factory is initialized they are empty, while they become simultaneously populated when running the factory.  Now instead `_instack ` is directly initialized by `__init__`  which means that, before running the factory, we have an intermediate situation in which `_instack` is full while `_outstack` is empty.
>   If you think this is ok, we can leave the current structure. Otherwise an alternative option could be to define a private a method `_update_instack()` and to call it only when the measured expectation values are saved into `_outstack`. I have no strong preference here, just to double checking we are aware of this change of "unwritten convention".

It's a good point and we should be consistent. I'll populate the `_instack` only when a `run` method is called. (This involves storing the scale factors and shots in the constructor, however.)


> * Shall we rename `iterate` to `run_classical` also in AdaptiveFactory? (To have a more uniform names).

I think having different names is actually good here since adaptive factories actually do iterate. 

Note that if we end up just adding a `run_batched` method (as I've done here), this could be done with the current factory structure by just defining `run_batched` in `BatchedFactory`. Then `iterate` would be a batched factory method.

> * I left two minor code suggestions about `kwargs`.

Thanks! I'll look into these, my current handling of kwargs is very hastily put together and should be made better.

File: mitiq/zne/inference/factories.py [hunk 1]
```diff
@@ -195,6 +98,176 @@ def get_expectation_values(self) -> np.ndarray:
...
+    ) -> "BatchedFactory":
+        """Computes the expectation values at each scale factor by calling
+        `batched_executor` on a list of all circuits to run.
+
+        Args:
+            qp: Quantum circuit to run.
+            batched_executor: Function which inputs a list of circuits and
+                outputs a list of expectation values.
+            scale_noise: Noise scaling function.
+            scale_noise: Noise scaling function.
```
Comment (nathanshammah) [inline mitiq/zne/inference/factories.py:197]:
```suggestion
            scale_noise: Noise scaling function.
```

**Comment by rmlarose** [comment]
Latest commit checks the executor signature and runs batched/sequential appropriately. See [here](https://github.com/unitaryfund/mitiq/blob/a928c11f8e6c6ca6960ea02ffdbbe7290ca72e3d/mitiq/zne/inference/factories.py#L265). Let me know what you think @karalekas @andreamari @willzeng.

Note that a function annotation is required with this method to do the batching. If there is no annotation, a "single executor" is assumed.

This can be spruced up a bit, but this is the basic idea. One can even try to get creative with [inspect](https://docs.python.org/3/library/inspect.html#inspect.getsourcelines) and auto generate a batched executor from a single executor, but let's keep it simple for now.

**Comment by andreamari** [comment]
> Latest commit checks the executor signature and runs batched/sequential appropriately. See [here](https://github.com/unitaryfund/mitiq/blob/a928c11f8e6c6ca6960ea02ffdbbe7290ca72e3d/mitiq/zne/inference/factories.py#L265), screenshot below. Let me know what you think @karalekas @andreamari @willzeng.
> 
> image
> 
> Note that a function annotation is required with this method to do the batching. If there is no annotation, a "single executor" is assumed.
> 
> This can be spruced up a bit, but this is the basic idea. One can even try to get creative with [inspect](https://docs.python.org/3/library/inspect.html#inspect.getsourcelines) and auto generate a batched executor from a single executor, but let's keep it simple for now.

Using the annotation seems a reasonable solution to me: if you are a user really interested in batching, then you only need to remember to annotate the executor, otherwise nothing is broken.

If we use this approach, I would also support the idea of internally building a batched executor from a single-circuit executor instead of having two methods (`run` and `run_batched`) with very similar code. But this is a minor aspect.

**Comment by rmlarose** [comment]
Great, thanks @andreamari. Unless anyone is opposed, I will tidy up this PR using the annotated executor approach.

We can open a separate issue in the future on generating a batched executor from a single executor, if even possible to do so in general.

**Comment by willzeng** [comment]
Nice find @rmlarose! I'm glad Python can do this fairly cleanly.

By the way, it wasn't obvious to me that checking the return type is better than checking the input type, but it is! In fact this method would allow someone to write a executor with type `QProgram --> List[float]` that broke up some number of shots across several quantum computers in parallel, pretty snazzy 🎩

**Comment by rmlarose** [comment]
Superseded by #378.

**Pull request closed without merging**